### PR TITLE
Fix warnings from the spring boot starter

### PR DESCRIPTION
**Pull request by laurit**
Resolves https://github.com/open-telemetry/opentelemetry-java-instrumentation/issues/9988
As far as I understand these warnings are printed for beans that are dependencies for one of our BeanPostProcessors. This pr attempts to resolve this be declaring the methods that return a BeanPostProcessor as static to avoid having dependency on the bean that contains the method and instead of `OpenTelemetry` inject a `Supplier` to avoid having a direct dependency on `OpenTelemetry` bean and its dependencies.

### CODE REVIEW

File: instrumentation/spring/spring-boot-autoconfigure/src/main/java/io/opentelemetry/instrumentation/spring/autoconfigure/OpenTelemetryAutoConfiguration.java [hunk 1]
```diff
@@ -169,4 +172,10 @@ public OpenTelemetry openTelemetry() {
       return OpenTelemetry.noop();
     }
   }
+
+  @Bean
+  @Role(BeanDefinition.ROLE_INFRASTRUCTURE)
+  public static OpenTelemetrySupplier openTelemetrySupplier(BeanFactory beanFactory) {
```
Comment (jeanbisutti) [inline instrumentation/spring/spring-boot-autoconfigure/src/main/java/io/opentelemetry/instrumentation/spring/autoconfigure/OpenTelemetryAutoConfiguration.java:178]: Could you please add a comment about the why?

File: instrumentation/spring/spring-boot-autoconfigure/src/main/java/io/opentelemetry/instrumentation/spring/autoconfigure/OpenTelemetrySupplier.java [hunk 2]
```diff
@@ -0,0 +1,12 @@
...
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package io.opentelemetry.instrumentation.spring.autoconfigure;
+
+import io.opentelemetry.api.OpenTelemetry;
+import java.util.function.Supplier;
+
+/** To inject an OpenTelemetry into bean post processors */
+public interface OpenTelemetrySupplier extends Supplier<OpenTelemetry> {}
```
Comment (jeanbisutti) [inline instrumentation/spring/spring-boot-autoconfigure/src/main/java/io/opentelemetry/instrumentation/spring/autoconfigure/OpenTelemetrySupplier.java:12]: Is the addition of an interface really needed?

Comment (jeanbisutti) [reply instrumentation/spring/spring-boot-autoconfigure/src/main/java/io/opentelemetry/instrumentation/spring/autoconfigure/OpenTelemetrySupplier.java:12]: Should the comment specify that the class is for internal use?

Comment (jeanbisutti) [reply instrumentation/spring/spring-boot-autoconfigure/src/main/java/io/opentelemetry/instrumentation/spring/autoconfigure/OpenTelemetrySupplier.java:12]: The interface name should perhaps mentioned that this interface is only for post processor usage. `OpentTelemerySupplierInPostProcessor`?

File: instrumentation/spring/spring-boot-autoconfigure/src/main/java/io/opentelemetry/instrumentation/spring/autoconfigure/OpenTelemetryAutoConfiguration.java (hunk 1 above)
Comment (laurit) [reply instrumentation/spring/spring-boot-autoconfigure/src/main/java/io/opentelemetry/instrumentation/spring/autoconfigure/OpenTelemetryAutoConfiguration.java:181]: comment added

File: instrumentation/spring/spring-boot-autoconfigure/src/main/java/io/opentelemetry/instrumentation/spring/autoconfigure/OpenTelemetrySupplier.java (hunk 2 above)
Comment (laurit) [reply instrumentation/spring/spring-boot-autoconfigure/src/main/java/io/opentelemetry/instrumentation/spring/autoconfigure/OpenTelemetrySupplier.java]: I moved it to internal package, but left the name as is. 

> Is the addition of an interface really needed?

Do you have a better idea how to do this? To me using a separate interface felt like the simplest approach to ensure that there is only 1 bean that matches.

Comment (jeanbisutti) [reply instrumentation/spring/spring-boot-autoconfigure/src/main/java/io/opentelemetry/instrumentation/spring/autoconfigure/OpenTelemetrySupplier.java]: I agree. By declaring a bean of type `Supplier<OpenTelemetry>`, the user could also declare in his code this type of bean and the application will fail to start.

**Review by jeanbisutti — APPROVED** [review]

**Review by trask — APPROVED** [review]

**Pull request merged**